### DISCUSSION TITLE
how to get value from a json array of parameters?

### DISCUSSION THREAD

**Discussion by StefXinet** [nlohmann/json#4290] [Q&A] [answered]
Hello,
I have a JSON file formatted like this:
```json
{
   "json_info":{
      "file_ver":1,
      "target":"device1"
   },
   "map":[
      {
         "paramA":0,
         "paramB":49,
         "name":"setting1"
      },
      {
         "paramA":56,
         "paramB":-25,
         "name":"setting2"
      },
      {
         "paramA":7,
         "paramB":34,
         "name":"setting3"
      }
   ]
}
```
And I'm reading it like this:
```C
int main (){
std::string file_name;
/* some code here to get the right filename and putting it into "file_name" var */
std::ifstream file( file_name );
json jfile = json::parse( file );				// json object with the full JSON file content
json jmap = jfile["map"];					// json object with the map ARRAY
}
```
I can correctly read datas from "json_info" object like this:
```C
string target = jfile["json_info"]["target"];
```
but I really cant understand how to read all the setting* from the "map" array.
I started putting the "map" into a json object, and basically wat I should do is to look for an array element which has the right "name" and read its "paramA" and "paramB" datas.
```C
json jmap = jfile["map"];							// json object with the map ARRAY
int arraysize = jmap.size();						// JSON map array elements, this works well
for( auto it = jmap.begin(); it != jmap.end(); ++it ) {	// loops on every array element
	json jitem = it.value();
	TRACE("name: %s; value: %d\n", jitem["name"], jitem["paramA"]);			// does not work
	TRACE("name: %s; value: %d\n", jitem.find("name"), jitem.find("paramA"));	// does not work
}
```
basically using both  jitem["name"] or jitem.find("name") i cant find the right element value. Could you help me?
Thanks!

**Comment by gregmarr** [comment, marked as answer]
> basically using both jitem["name"] or jitem.find("name") i cant find the right element value. Could you help me?

You found the right element, you now just need to extract the data as the right type.

	TRACE("name: %s; value: %d\n", jitem["name"], jitem["paramA"]);			// does not work

This doesn't work because you are passing an object of type `json` with format strings of `%s` and `%d` which won't work.  The following will work:

	TRACE("name: %s; value: %d\n", jitem["name"].get<std::string>().c_str(), jitem["paramA"].get<int>());

or better

    std::cout << "name: " << jitem["name"].get<std::string>() << " value: " << jitem["paramA"].get<int>() << "\n";

**Comment by StefXinet** [comment]
You are right, thanks!
I already found the right element but I get stucked in debugging!
